AWS Glue 4.0 - NoSuchMethodError

Hi,

thanks for the great work!

I'm trying to run the latest agent (spark-3.3-spline-agent-bundle_2.12-1.0.4) with Glue 4.0 on AWS and a pyspark script, which fails in an exception. I was hoping maybe you could give me a hint what to look out for.

I've tested my code with the agent against "vanilla" Spark 3.3.* which works perfectly fine - obviously it cannot be exactly the same code because it is not running as an AWS managed service. Glue 4.0 should use Spark 3.3.0 as stated here: https://docs.aws.amazon.com/glue/latest/dg/release-notes.html

The same code is also working fine with Glue 3 (spark-3.1-spline-agent-bundle_2.12-1.0.4) and Glue 2 (spark-2.2-spline-agent-bundle_2.11-1.0.4).

I tried to create a minimal example for the problem:

Pyspark script:
``` Python
import dataclasses
import sys
from dataclasses import dataclass

from awsglue.context import GlueContext
from awsglue.utils import getResolvedOptions
from pyspark import SparkConf
from pyspark.context import SparkContext


@dataclass(frozen=True)
class Arguments:
    """Stores and defines the program arguments."""

    source_database: str
    source_table: str
    target_database: str
    target_table: str
    s3_output_path: str
    kms_key_arn: str


def parse_arguments() -> Arguments:
    """Validate and parse the script arguments."""
    field_names = [field.name for field in dataclasses.fields(Arguments)]
    all_arguments = getResolvedOptions(sys.argv, field_names)
    filtered_arguments = {key: value for key, value in all_arguments.items() if key in field_names}
    return Arguments(**filtered_arguments)


def main() -> None:
    """Contain the primary logic of the job."""
    arguments = parse_arguments()
    spark_settings = {
        "spark.spline.mode": "ENABLED",
        "spark.sql.queryExecutionListeners": "za.co.absa.spline.harvester.listener.SplineQueryExecutionListener",
        "spark.spline.lineageDispatcher": "console",
        "spark.spline.lineageDispatcher.console.className": "za.co.absa.spline.harvester.dispatcher.ConsoleLineageDispatcher",
    }
    spark_config = SparkConf().setAll([(key, value) for key, value in spark_settings.items()])
    spark_context = SparkContext(conf=spark_config)
    spark_context._jsc.hadoopConfiguration().set("fs.s3.enableServerSideEncryption", "true")
    spark_context._jsc.hadoopConfiguration().set("fs.s3.serverSideEncryption.kms.keyId", arguments.kms_key_arn)
    spark_context._jsc.hadoopConfiguration().set("fs.s3.canned.acl", "BucketOwnerFullControl")
    glue_context = GlueContext(spark_context)
    glue_context.setConf("spark.sql.parquet.writeLegacyFormat", "true")
    glue_context.setConf("spark.hadoop.fs.s3.maxRetries", "50")

    spark_session = glue_context.spark_session
    query = f"SELECT * FROM {arguments.source_database}.{arguments.source_table}"
    print(query)  # noqa: T201

    data_frame = spark_session.sql(query)
    data_frame = data_frame.drop("Foo").withColumnRenamed("Bar", "Baz")

    print(f"count: {data_frame.count()}")  # noqa: T201

    data_frame.write.option("path", arguments.s3_output_path).mode("overwrite").saveAsTable(
        f"{arguments.target_database}.{arguments.target_table}", format="parquet"
    )
    print("finished")  # noqa: T201


if __name__ == "__main__":
    main()

```
Glue job config:
``` Json
{
  "jobConfig": {
    "name": "test_spline_agent_4",
    "description": "",
    "role": "arn:aws:iam::***",
    "command": "glueetl",
    "version": "4.0",
    "workerType": "Standard",
    "numberOfWorkers": 2,
    "maxCapacity": 2,
    "maxRetries": 0,
    "timeout": 2880,
    "maxConcurrentRuns": 10,
    "security": "none",
    "scriptName": "test_spline_agent_4.py",
    "scriptLocation": "s3://***/",
    "language": "python-3",
    "jobParameters": [
      {
        "key": "--kms_key_arn",
        "value": "arn:aws:kms:***",
        "existing": false
      },
     {
        "key": "--s3_output_path",
        "value": "s3://***",
        "existing": false
      },
      {
        "key": "--source_database",
        "value": "***",
        "existing": false
      },
      {
        "key": "--source_table",
        "value": "***",
        "existing": false
      },
      {
        "key": "--target_database",
        "value": "***",
        "existing": false
      },
      {
        "key": "--target_table",
        "value": "***",
        "existing": false
      }
   ],
    "tags": [],
    "jobMode": "DEVELOPER_MODE",
    "developerMode": false,
    "connectionsList": [],
    "temporaryDirectory": "s3://***/tmp",
    "logging": true,
    "glueHiveMetastore": true,
    "metrics": false,
    "dependentPath": "s3://***/spark-3.3-spline-agent-bundle_2.12-1.0.4.jar,s3://***/snakeyaml_1.33.jar",
    "bookmark": "job-bookmark-disable",
    "flexExecution": false,
    "minFlexWorkers": null,
    "pythonPath": null
  },
  "hasBeenSaved": false,
  "script": "***"
}
```

Log / Error:
```
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Initializing Spline Agent...
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Spline Version: 1.0.4 (rev. 4266fca)
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Init Type: AUTO (codeless)
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Init Mode: ENABLED
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Lineage Dispatcher: Console
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Post-Processing Filter: Password replace
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Ignore-Write Detection Strategy: Write metrics
...
INFO [Thread-7] harvester.SparkLineageInitializer (Logging.scala:logInfo(61)): Spline successfully initialized. Spark Lineage tracking is ENABLED.
...
ERROR [spark-listener-group-shared] util.Utils (Logging.scala:logError(98)): throw uncaught fatal error in thread spark-listener-group-shared
java.lang.ExceptionInInitializerError: null
   at za.co.absa.spline.harvester.HashBasedUUIDGenerator.nextId(idGenerators.scala:57) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.DataTypeIdGenerator.nextId(idGenerators.scala:83) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.converter.DataTypeConverter.convert(DataTypeConverter.scala:39) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1$$anon$2.za$co$absa$commons$lang$CachingConverter$$super$convert(SplineAgent.scala:76) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:206) ~[scala-library.jar:?]
   at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:203) ~[scala-library.jar:?]
   at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:80) ~[scala-library.jar:?]
   at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1$$anon$2.convert(SplineAgent.scala:76) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1$$anon$2.convert(SplineAgent.scala:76) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.converter.DataTypeConverter.convert(DataTypeConverter.scala:43) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.converter.AttributeConverter.convert(AttributeConverter.scala:42) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$1.za$co$absa$commons$lang$CachingConverter$$super$convert(PlanOperationNodeBuilder.scala:36) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:206) ~[scala-library.jar:?]
   at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:203) ~[scala-library.jar:?]
   at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:80) ~[scala-library.jar:?]
   at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$1.convert(PlanOperationNodeBuilder.scala:36) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder.$anonfun$outputAttributes$1(PlanOperationNodeBuilder.scala:66) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233) ~[scala-library.jar:?]
   at scala.collection.immutable.List.foreach(List.scala:388) ~[scala-library.jar:?]
   at scala.collection.TraversableLike.map(TraversableLike.scala:233) ~[scala-library.jar:?]
   at scala.collection.TraversableLike.map$(TraversableLike.scala:226) ~[scala-library.jar:?]
   at scala.collection.immutable.List.map(List.scala:294) ~[scala-library.jar:?]
   at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder.outputAttributes(PlanOperationNodeBuilder.scala:66) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder.outputAttributes$(PlanOperationNodeBuilder.scala:65) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.read.ReadNodeBuilder.outputAttributes$lzycompute(ReadNodeBuilder.scala:28) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.read.ReadNodeBuilder.outputAttributes(ReadNodeBuilder.scala:28) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.read.ReadNodeBuilder.build(ReadNodeBuilder.scala:42) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.builder.plan.read.ReadNodeBuilder.build(ReadNodeBuilder.scala:28) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.LineageHarvester.$anonfun$harvest$6(LineageHarvester.scala:72) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233) ~[scala-library.jar:?]
   at scala.collection.immutable.List.foreach(List.scala:388) ~[scala-library.jar:?]
   at scala.collection.TraversableLike.map(TraversableLike.scala:233) ~[scala-library.jar:?]
   at scala.collection.TraversableLike.map$(TraversableLike.scala:226) ~[scala-library.jar:?]
   at scala.collection.immutable.List.map(List.scala:294) ~[scala-library.jar:?]
   at za.co.absa.spline.harvester.LineageHarvester.$anonfun$harvest$4(LineageHarvester.scala:72) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.Option.flatMap(Option.scala:171) ~[scala-library.jar:?]
   at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:65) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1.$anonfun$handle$1(SplineAgent.scala:91) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1.withErrorHandling(SplineAgent.scala:100) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.agent.SplineAgent$$anon$1.handle(SplineAgent.scala:72) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.listener.QueryExecutionListenerDelegate.onSuccess(QueryExecutionListenerDelegate.scala:28) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1(SplineQueryExecutionListener.scala:41) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1$adapted(SplineQueryExecutionListener.scala:41) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at scala.Option.foreach(Option.scala:257) ~[scala-library.jar:?]
   at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:41) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:165) ~[spark-sql_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:135) ~[spark-sql_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:135) ~[spark-sql_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:147) ~[spark-sql_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:100) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:117) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:101) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:12) ~[scala-library.jar:?]
   at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58) ~[scala-library.jar:?]
   at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1447) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
   at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96) ~[spark-core_2.12-3.3.0-amzn-1.jar:3.3.0-amzn-1]
Caused by: scala.tools.reflect.ToolBoxError: reflective compilation has failed: cannot initialize the compiler due to java.lang.NoSuchMethodError: scala.tools.reflect.package$$anon$4.INFO()Lscala/reflect/internal/Reporter$Severity;
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.liftedTree1$1(ToolBoxFactory.scala:360) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.compiler$lzycompute(ToolBoxFactory.scala:346) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.compiler(ToolBoxFactory.scala:345) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$.apply(ToolBoxFactory.scala:372) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl.parse(ToolBoxFactory.scala:429) ~[scala-compiler-2.12.15.jar:?]
   at za.co.absa.spline.harvester.json.HarvesterJsonSerDe$.<init>(HarvesterJsonSerDe.scala:49) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.json.HarvesterJsonSerDe$.<clinit>(HarvesterJsonSerDe.scala) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   ... 71 more
Caused by: java.lang.NoSuchMethodError: scala.tools.reflect.package$$anon$4.INFO()Lscala/reflect/internal/Reporter$Severity;
   at scala.tools.reflect.package$$anon$4.<init>(package.scala:85) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.package$.frontEndToReporter(package.scala:77) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.liftedTree1$1(ToolBoxFactory.scala:350) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.compiler$lzycompute(ToolBoxFactory.scala:346) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$api$.compiler(ToolBoxFactory.scala:345) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$.apply(ToolBoxFactory.scala:372) ~[scala-compiler-2.12.15.jar:?]
   at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl.parse(ToolBoxFactory.scala:429) ~[scala-compiler-2.12.15.jar:?]
   at za.co.absa.spline.harvester.json.HarvesterJsonSerDe$.<init>(HarvesterJsonSerDe.scala:49) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   at za.co.absa.spline.harvester.json.HarvesterJsonSerDe$.<clinit>(HarvesterJsonSerDe.scala) ~[spark-3.3-spline-agent-bundle_2.12-1.0.4.jar:?]
   ... 71 more
```

The Glue job finishes successfully only the agent fails.

I've noticed that Glue 4 is missing the snake yaml library (https://mvnrepository.com/artifact/org.yaml/snakeyaml/1.33) that's why I added it manually to the path.

I thought the error might be related to https://github.com/AbsaOSS/spline-spark-agent/issues/382. Therefore I checked the Scala version Glue 4 is using. To do so I've added a plugin as descibed here: https://github.com/AbsaOSS/spline-spark-agent#plugin-api, which prints ```scala.util.Properties.versionNumberString```-> `2.12.7`, which should be okay?! Just to double check I also recompiled the agent with 2.12.7, but the result was the same error.

I also tested the versions spark-3.3-spline-agent-bundle_2.12-0.7.10 and spark-3.3-spline-agent-bundle_2.12-0.7.13, but it resulted in the same error.

But now I'm out of ideas... Could you point me towards something I could check out?

Best regards,
Philip

--- a/bundle-2.2/pom.xml
+++ b/bundle-2.2/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-32.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-2.3/pom.xml
+++ b/bundle-2.3/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-32.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-2.4/pom.xml
+++ b/bundle-2.4/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-35.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-3.0/pom.xml
+++ b/bundle-3.0/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-36.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-3.1/pom.xml
+++ b/bundle-3.1/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-37-M5.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-3.2/pom.xml
+++ b/bundle-3.2/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-37-M11.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/bundle-3.3/pom.xml
+++ b/bundle-3.3/pom.xml
@@ -33,11 +33,6 @@
             <artifactId>agent-core_${scala.binary.version}</artifactId>
             <version>${project.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <version>${json4s-37-M11.version}</version>
-        </dependency>
     </dependencies>
 
     <dependencyManagement>

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -45,11 +45,6 @@
 
     <dependencies>
         <dependency>
-            <groupId>org.scala-lang</groupId>
-            <artifactId>scala-compiler</artifactId>
-            <version>${scala.version}</version>
-        </dependency>
-        <dependency>
             <groupId>za.co.absa.commons</groupId>
             <artifactId>commons_${scala.binary.version}</artifactId>
             <version>1.3.3</version>
@@ -89,6 +84,14 @@
             <groupId>org.yaml</groupId>
             <artifactId>snakeyaml</artifactId>
             <version>1.33</version>
+        </dependency>
+        <dependency>
+            <groupId>org.json4s</groupId>
+            <artifactId>json4s-native_${scala.binary.version}</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.json4s</groupId>
+            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
         </dependency>
 
         <!-- optional dependencies (for plugins) -->
@@ -137,14 +140,6 @@
             <groupId>za.co.absa.cobrix</groupId>
             <artifactId>spark-cobol_${scala.binary.version}</artifactId>
             <optional>true</optional>
-        </dependency>
-
-        <!-- test scope -->
-
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <scope>test</scope>
         </dependency>
     </dependencies>
 
@@ -258,6 +253,37 @@
             <plugin>
                 <groupId>com.github.cerveada</groupId>
                 <artifactId>scalatest-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <createDependencyReducedPom>false</createDependencyReducedPom>
+                    <minimizeJar>true</minimizeJar>
+                    <artifactSet>
+                        <includes>
+                            <include>org.json4s:json4s-native_${scala.binary.version}</include>
+                            <include>org.json4s:json4s-core_${scala.binary.version}</include>
+                            <include>org.json4s:json4s-ast_${scala.binary.version}</include>
+                            <include>org.json4s:json4s-scalap_${scala.binary.version}</include>
+                            <include>org.json4s:json4s-ext_${scala.binary.version}</include>
+                        </includes>
+                    </artifactSet>
+                    <relocations>
+                        <relocation>
+                            <pattern>org.json4s</pattern>
+                            <shadedPattern>za.co.absa.spline.shaded.org.json4s</shadedPattern>
+                        </relocation>
+                    </relocations>
+                </configuration>
             </plugin>
         </plugins>
     </build>

--- a/core/src/main/scala/za/co/absa/spline/harvester/json/HarvesterJsonSerDe.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/json/HarvesterJsonSerDe.scala
@@ -16,37 +16,28 @@
 
 package za.co.absa.spline.harvester.json
 
-import org.json4s.JValue
-import org.json4s.jackson.JsonMethods
-import za.co.absa.commons.json.AbstractJsonSerDe
-import za.co.absa.commons.json.format.JavaTypesSupport
-import za.co.absa.commons.reflect.ReflectionUtils
-
-import scala.reflect.ClassTag
-import scala.reflect.runtime.universe._
-import scala.tools.reflect.ToolBox
+import org.json4s.Extraction.decompose
+import org.json4s.ext.JavaTypesSerializers
+import org.json4s.native.JsonMethods.{compact, parse, pretty, render}
+import org.json4s.{Formats, StringInput}
 
 object HarvesterJsonSerDe {
 
-  private def classNameOf[A: ClassTag] = {
-    implicitly[ClassTag[A]].runtimeClass.getCanonicalName
-  }
+  object impl {
+    private implicit val formats: Formats =
+      ShortTypeHintForSpline03ModelSupport.formats ++
+        JavaTypesSerializers.all
 
-  private val JsonMethods = classNameOf[JsonMethods]
-  private val JavaTypesSupport = classNameOf[JavaTypesSupport]
-  private val ShortTypeHintForSpline03ModelSupport = classNameOf[ShortTypeHintForSpline03ModelSupport]
-  private val JValue = classNameOf[JValue]
-  private val AbstractJsonSerDe = classNameOf[AbstractJsonSerDe[_]]
+    implicit class EntityToJson[B <: AnyRef](entity: B) {
+      def toJson: String = compact(render(decompose(entity)(formats)))
+    }
 
-  // This delays the compilation (to bytecode) of that piece of code at runtime.
-  // Commons are build against json4s 3.5.5, spark 2.4 usually provides json4s 3.5.3 and these are not binary compatible!
-  val impl: AbstractJsonSerDe[JValue] = {
-    ReflectionUtils.compile(runtimeMirror(getClass.getClassLoader).mkToolBox().parse(
-      s"""
-        new $AbstractJsonSerDe[$JValue]
-          with $JsonMethods
-          with $ShortTypeHintForSpline03ModelSupport
-          with $JavaTypesSupport
-      """))(Map.empty)
+    implicit class JsonToEntity(json: String) {
+      def fromJson[B: Manifest]: B = parseJson().extract(formats, implicitly[Manifest[B]])
+
+      def asPrettyJson: String = pretty(render(parseJson()))
+
+      private def parseJson() = parse(StringInput(json), formats.wantsBigDecimal, formats.wantsBigInt)
+    }
   }
 }

--- a/core/src/main/scala/za/co/absa/spline/harvester/json/ShortTypeHintForSpline03ModelSupport.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/json/ShortTypeHintForSpline03ModelSupport.scala
@@ -18,75 +18,27 @@ package za.co.absa.spline.harvester.json
 
 import org.apache.commons.lang3.StringUtils._
 import org.json4s.{DefaultFormats, Formats, TypeHints}
-import za.co.absa.commons.json.format.FormatsBuilder
 import za.co.absa.commons.reflect.ReflectionUtils._
-import za.co.absa.spline.harvester.json.ShortTypeHintForSpline03ModelSupport._
-import za.co.absa.spline.model
+import za.co.absa.spline.model.dt.DataType
 
-import scala.reflect.runtime.universe._
-
-trait ShortTypeHintForSpline03ModelSupport extends FormatsBuilder {
-  override protected def formats: Formats = createFormats(Map(
-    "typeHintFieldName" -> "_typeHint",
-    "typeHintClasses" -> directSubClassesOf[model.dt.DataType],
-    "dateFormatterFn" -> DefaultFormats.losslessDate.get _))
-}
+import java.text.SimpleDateFormat
 
 object ShortTypeHintForSpline03ModelSupport {
+  val formats: Formats = new DefaultFormats {
+    override def dateFormatter: SimpleDateFormat = DefaultFormats.losslessDate.get
 
-  private val createFormats =
-    if (isJson4sVerAtLeast37) createFormats37
-    else createFormats32
+    override val typeHintFieldName: String = "_typeHint"
 
-  private def isJson4sVerAtLeast37: Boolean =
-    classOf[TypeHints].getMethods.exists(_.getName == "typeHintFieldName")
+    override val typeHints: TypeHints = new TypeHints {
+      private val classes: Seq[Class[_ <: DataType]] = directSubClassesOf[DataType]
+      override val hints: List[Class[_]] = classes.toList
 
-  abstract class AbstractSplineShortTypeHints(classes: Seq[Class[_]]) extends TypeHints {
-    override val hints: List[Class[_]] = classes.toList
+      override def hintFor(clazz: Class[_]): String = {
+        val className = clazz.getName
+        className.substring(1 + lastOrdinalIndexOf(className, ".", 2))
+      }
 
-    def classFor(hint: String, a: Class[_]): Option[Class[_]] = classFor(hint)
-
-    def classFor(hint: String): Option[Class[_]] = classes find (hintForAsString(_) == hint)
-
-    // TypeHints.hintFor() has a different return type in Json4s >= 3.7
-    // So we should not use it directly, and use this method instead.
-    protected def hintForAsString(clazz: Class[_]): String = {
-      val className = clazz.getName
-      className.substring(1 + lastOrdinalIndexOf(className, ".", 2))
+      def classFor(hint: String): Option[Class[_]] = classes find (hintFor(_) == hint)
     }
   }
-
-  private def createFormats32: Map[String, Any] => Formats = compile[Formats](
-    q"""
-      import java.text.SimpleDateFormat
-      import org.json4s.{DefaultFormats, TypeHints}
-      import za.co.absa.spline.harvester.json.ShortTypeHintForSpline03ModelSupport.AbstractSplineShortTypeHints
-
-      new DefaultFormats {
-        override def dateFormatter = args[() => SimpleDateFormat]("dateFormatterFn")()
-        override val typeHintFieldName: String = args("typeHintFieldName")
-        override val typeHints: TypeHints = {
-          new AbstractSplineShortTypeHints(args("typeHintClasses")) {
-            override def hintFor(clazz: Class[_]): String = super.hintForAsString(clazz)
-          }
-        }
-      }
-    """)
-
-  private def createFormats37: Map[String, Any] => Formats = compile[Formats](
-    q"""
-      import java.text.SimpleDateFormat
-      import org.json4s.{DefaultFormats, TypeHints}
-      import za.co.absa.spline.harvester.json.ShortTypeHintForSpline03ModelSupport.AbstractSplineShortTypeHints
-
-      new DefaultFormats {
-        override def dateFormatter = args[() => SimpleDateFormat]("dateFormatterFn")()
-        override val typeHints: TypeHints = {
-          new AbstractSplineShortTypeHints(args("typeHintClasses")) {
-            override def hintFor(clazz: Class[_]): Option[String] = Option(super.hintForAsString(clazz))
-            override val typeHintFieldName: String = args("typeHintFieldName")
-          }
-        }
-      }
-    """)
 }

--- a/examples/pom.xml
+++ b/examples/pom.xml
@@ -44,10 +44,6 @@
             <artifactId>spark-sql_${scala.binary.version}</artifactId>
             <version>${spark.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-        </dependency>
 
         <!-- runtime dependencies -->
 

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -52,11 +52,6 @@
         <!-- Test scope -->
 
         <dependency>
-            <groupId>org.json4s</groupId>
-            <artifactId>json4s-ext_${scala.binary.version}</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
             <groupId>org.apache.spark</groupId>
             <artifactId>spark-hive_${scala.binary.version}</artifactId>
             <scope>test</scope>

--- a/integration-tests/src/test/scala/za/co/absa/spline/harvester/SparkLineageInitializerSpec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/harvester/SparkLineageInitializerSpec.scala
@@ -28,7 +28,6 @@ import org.scalatest.matchers.should.Matchers
 import org.scalatest.{Assertion, BeforeAndAfter, Succeeded}
 import org.scalatestplus.mockito.MockitoSugar
 import za.co.absa.commons.io.TempFile
-import za.co.absa.commons.json.DefaultJacksonJsonSerDe
 import za.co.absa.commons.scalatest.ConditionalTestTags._
 import za.co.absa.commons.version.Version._
 import za.co.absa.spline.agent.AgentConfig.ConfProperty
@@ -220,7 +219,7 @@ object SparkLineageInitializerSpec {
     override def send(event: ExecutionEvent): Unit = MockLineageDispatcher.theMock.send(event)
   }
 
-  private object MockLineageDispatcher extends MockitoSugar with DefaultJacksonJsonSerDe {
+  private object MockLineageDispatcher extends MockitoSugar {
     private val theMock: LineageDispatcher = mock[LineageDispatcher]
     private[this] var throwableOnConstruction: Option[_ <: Throwable] = None
     private[this] var _instanceCount: Int = _

--- a/integration-tests/src/test/scala/za/co/absa/spline/harvester/dispatcher/HDFSLineageDispatcherSpec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/harvester/dispatcher/HDFSLineageDispatcherSpec.scala
@@ -21,9 +21,9 @@ import org.apache.spark.SPARK_VERSION
 import org.scalatest.flatspec.AsyncFlatSpec
 import org.scalatest.matchers.should.Matchers
 import za.co.absa.commons.io.TempDirectory
-import za.co.absa.commons.json.DefaultJacksonJsonSerDe
 import za.co.absa.commons.scalatest.ConditionalTestTags.ignoreIf
 import za.co.absa.commons.version.Version._
+import za.co.absa.spline.harvester.json.HarvesterJsonSerDe.impl._
 import za.co.absa.spline.test.fixture.SparkFixture
 import za.co.absa.spline.test.fixture.spline.SplineFixture
 
@@ -33,8 +33,7 @@ class HDFSLineageDispatcherSpec
   extends AsyncFlatSpec
     with Matchers
     with SparkFixture
-    with SplineFixture
-    with DefaultJacksonJsonSerDe {
+    with SplineFixture {
 
   behavior of "HDFSLineageDispatcher"
 

--- a/pom.xml
+++ b/pom.xml
@@ -290,7 +290,7 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-shade-plugin</artifactId>
-                    <version>3.2.4</version>
+                    <version>3.4.1</version>
                     <configuration>
                         <createDependencyReducedPom>false</createDependencyReducedPom>
                         <minimizeJar>true</minimizeJar>

